framework/calico_framework.py

```diff
@@ -0,0 +1,661 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import os
+import sys
+import subprocess
+import re
+import json
+import time
+import threading
+import math
+from itertools import izip_longest
+from random import randint
+import mesos.interface
+from mesos.interface import mesos_pb2
+import mesos.native
+from calico_utils import _setup_logging
+from framework.tasks import SleepTask, PingTask, NetcatListenTask, NetcatSendTask
+from framework.constants import LOGFILE, TASK_CPUS, TASK_MEM, UNFINISHED_TASK_STATES, BAD_TASK_STATES, TEST_TIMEOUT
+
+_log = _setup_logging(LOGFILE)
+NEXT_AVAILABLE_TASK_ID = 0
+
+class TestCase(object):
+    def __init__(self, tasks, name):
+        self.name = name
+        """
+        Nicename to identify this test.
+        """
+
+        self.tasks = tasks
+        """
+        Tasks for this test.
+        """
+
+        # Add a backwards reference from each Task to its parent test
+        for task in tasks:
+            task.test = self
+
+        self.timeout = None
+        """
+        If None, this test is not being tracked for timeout.
+
+        Otherwise, this number will be seconds-since-epoch of the last checkpoint.
+
+        For an unstarted test, this is set if/when
+        this test refuses to start while no other tests are running.
+        If time expires in this state, that means the mesos cluster was
+        unable to provide enough offers to launch this test when no other tests were running.
+
+        For a started test, this value is reset at each task update, and represents the last time
+        this Test heard from one of its tasks.
+        """
+
+        self.additional_info = None
+        """
+        Test-wide information which will be provided during the test output.
+        Useful for explaining a test-scope failure (vs. a task failure).
+        """
+
+        self.killed = False
+        """
+        Flag, when set to True, will indicate that this test is being shutdown.
+        """
+
+    def __repr__(self):
+        s_repr = "Test(%s" % self.name
+        if self.timeout != None:
+            s_repr += (", timeout=%d" % (time.time() - self.timeout))
+        s_repr += ")"
+
+        return s_repr
+
+    @property
+    def complete(self):
+        """
+        True if the test is no longer running. This means killed tests too
+        """
+        return self.killed or self.tasks_with_state(UNFINISHED_TASK_STATES) == []
+
+    def can_run_on(self, offers):
+        """
+        Returns true if the test can run on the offers passed in.
+        Offers passed in will be ignored if they are already reserved.
+        This function also reserves offers and plans tasks as it goes along.
+        It will rollback reservations if it realizes it can't finish planning all tasks.
+        """
+        offers = [offer for offer in offers if offer.reserved_by in [None, self]]
```

Comment:
A little confused by this filtering.  Why `self`?  Don't we only call this on tests that haven't yet launched?